### DISCUSSION TITLE
Serverpod's file URL format doesn't work with Flutter network image caching

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#1227] [Q&A] [answered]
I have the following three URLs, for files I uploaded to serverpod:

```console
http://10.0.2.2:8080/serverpod_cloud_storage?method=file&path=p%2F2448436A3A79t497li13.jpg
http://10.0.2.2:8080/serverpod_cloud_storage?method=file&path=p%2F26461212779153305U47.jpg
http://10.0.2.2:8080/serverpod_cloud_storage?method=file&path=p%2F9c0o79inkU87Pxu5992w.jpg
```

The first file (ending in `i13`) is cached fine by Flutter's HTTP client, when using `Network.image`. The second and third (`U47` and `92w`) are not -- they are fetched from the server every time their `Network.image` is rebuilt. This makes for a slow UI and a lot of unnecessary network traffic.

I believe the issue is that all three share a common root before the `?` character, and only the first file with a given path (before the query string) can be cached.

The URL format needs to use path syntax, not query string syntax, for caching to work properly. e.g.:

```console
http://10.0.2.2:8080/serverpod_cloud_storage/file/p/2448436A3A79t497li13.jpg
http://10.0.2.2:8080/serverpod_cloud_storage/file/p/26461212779153305U47.jpg
http://10.0.2.2:8080/serverpod_cloud_storage/file/p/9c0o79inkU87Pxu5992w.jpg
```

This would apply to the URL syntax of all of the cloud storage providers.

**Comment by vlidholt** [comment, marked as answer]
It's very much recommended that you use a cloud service, such as S3 or Google Cloud for uploading files. We added the database storage to get people quickly up and running, but for production apps you should use a more suited service than Postgres for this.

**Reply by lukehutch** [reply]
Obviously I'm not going to be serving millions of images from Postgres in production! However, debugging is so much easier when developing locally, and with current state of things, the UI is janky whenever images are served from Postgres. I need the local debugging experience to be as close to the deployed production experience as possible, including for caching behavior.

**Reply by vlidholt** [reply]
Good point! We'll try to fix this for 2.0 as it may be a breaking change. It's relatively easy to set up S3 with local development too, so that may still be an option in the mean time.

**Comment by lukehutch** [comment]
I was wrong about the cause of this. It turns out that Flutter's image cache has a fixed size of 100MB (for the *uncompressed size* of images, yikes), and I was exceeding that, which caused images to get evicted from the cache.

**Reply by lukehutch** [reply]
For the record, in case anyone else runs into this issue: the correct solution is to use the `cached_network_image` Flutter plugin, which uses a disk cache for compressed images, rather than a RAM cache for uncompressed images (like `Image.network` etc.).
